# Transcriptome_Curation

 This Git repository contains the whole pipeline used in my master's thesis to assemble a Richter 110 grapevine rootstock transcriptome from the genome published by Minio et al. (2022), using 98 RNA-seq samples of R110 root under different  normal and water and saline stress conditions. This workflow can be used to find new genes and 
improve the quality of transcriptomes from other non-model species coming from bioinformatic predictions using RNA-seq data.

## Background

The grapevine (*Vitis vinifera L.*) is one of the most important crops worldwide due to its relevance in the economic, environmental and social spheres of many countries. Since the 19th century, cultivated vines have been grafted onto roots of closely related Vitis species with desirable characteristics of resistance to the environment, called rootstocks. These improve the productivity and quality of the grape, allowing the plant to overcome biotic and abiotic stress conditions, as well as influencing the development of the stem and fruit. Most rootstocks are currently the result of crossing different vine species, allowing the combination of two desirable genotypes, which leads to the high heterozygosity of these hybrid organisms. Despite its relevance for viticulture and the high future expectations of the use of rootstocks as sustainable practices in the face of climate change, there are hardly any genomic resources available for these organisms, making molecular or breeding research difficult. <br>

This paper provides the transcriptome of the Richter 110 rootstock (*V.berlandieri cv. Rességuier nº2 x V.rupestris cv. Martin*) assembled by haplotypes, using root RNA-seq data in different conditions and using the R110 genome as reference. generated by Minio et al. (2022) and the consequences of using the appropriate annotation for differential gene expression analysis are demonstrated. <br>



## Requeriments
### Tools
&bull; [AnnotationForge v1.40.2](https://bioconductor.org/packages/release/bioc/html/AnnotationForge.html) <br>
&bull; [BLAST v2.13.0](https://blast.ncbi.nlm.nih.gov/Blast.cgi) <br>
&bull; [DESeq2 v1.38.0.](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) <br>
&bull; [EggNOG-mapper v1.0.3](http://eggnog-mapper.embl.de/) <br>
&bull; [fastQC v0.11.8](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) <br>
&bull; [GenomicRanges v1.50.2](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) <br>
&bull; [ggplot2 v3.4.2](https://ggplot2.tidyverse.org/) <br>
&bull; [GOstats v2.64.0](https://bioconductor.org/packages/release/bioc/html/GOstats.html) <br>
&bull; [HISAT2 v2.2.1](http://daehwankimlab.github.io/hisat2/) <br>
&bull; [MultiQC v1.14](https://multiqc.info/) <br>
&bull; [Rsubread v2.12.0](https://bioconductor.org/packages/release/bioc/html/Rsubread.html) <br>
&bull; [rrgvo v1.10.0](https://bioconductor.org/packages/release/bioc/html/rrvgo.html) <br>
&bull; [SAMtools v1.17](http://www.htslib.org/) <br>
&bull; [StringTie v2.2.1](https://ccb.jhu.edu/software/stringtie/) <br>
&bull; [TransDecoder v5.5.0](https://github.com/TransDecoder/TransDecoder/wiki) <br>
&bull; [trimmomatic v.0.36](http://www.usadellab.org/cms/?page=trimmomatic) <br>



## Steps to follow








